kolibri/core/auth/utils/delete.py

```diff
@@ -0,0 +1,305 @@
+                    logger.debug(debug_msg.format(count, obj_name, name))
+                all_deletions.update({obj_name: all_deletions.get(obj_name, 0) + count})
+            if self.sleep is not None:
+                time.sleep(sleep)
+
+        return total_count, all_deletions
+
+
+def chunk(things, size):
+    """
+    Chunk generator
+
+    :type things: list
+    :type size: int
+    """
+    for i in range(0, len(things), size):
+        yield things[i : i + size]
+
+
+def _get_facility_dataset(dataset_id):
+    return FacilityDataset.objects.filter(id=dataset_id)
+
+
+def _get_certificates(dataset_id):
+    return (
+        Certificate.objects.filter(id=dataset_id)
+        .get_descendants(include_self=True)
+        .exclude(_private_key=None)
+    )
+
+
+def _get_users(dataset_id):
+    user_id_filter = Q(
+        user_id__in=FacilityUser.objects.filter(dataset_id=dataset_id).values_list(
+            "pk", flat=True
+        )
+    )
+    dataset_id_filter = Q(dataset_id=dataset_id)
+
+    return GroupDeletion(
+        "User models",
+        querysets=[
+            LogEntry.objects.filter(user_id_filter),
+            DevicePermissions.objects.filter(user_id_filter),
+            PingbackNotificationDismissed.objects.filter(user_id_filter),
+            Collection.objects.filter(Q(parent_id__isnull=True) & dataset_id_filter),
+            Role.objects.filter(dataset_id_filter),
+            Membership.objects.filter(dataset_id_filter),
+            Bookmark.objects.filter(dataset_id_filter),
+            FacilityUser.objects.filter(dataset_id_filter),
+        ],
+    )
+
+
+def _get_class_models(dataset_id):
+    dataset_id_filter = Q(dataset_id=dataset_id)
+    return GroupDeletion(
+        "Class models",
+        querysets=[
+            ExamAssignment.objects.filter(dataset_id_filter),
+            Exam.objects.filter(dataset_id_filter),
+            LessonAssignment.objects.filter(dataset_id_filter),
+            Lesson.objects.filter(dataset_id_filter),
+            AdHocGroup.objects.filter(dataset_id_filter),
+            LearnerGroup.objects.filter(dataset_id_filter),
+            Classroom.objects.filter(dataset_id_filter),
+        ],
+    )
+
+
+def _get_log_models(dataset_id):
+    dataset_id_filter = Q(dataset_id=dataset_id)
+    return GroupDeletion(
+        "Log models",
+        querysets=[
+            ContentSessionLog.objects.filter(dataset_id_filter),
+            ContentSummaryLog.objects.filter(dataset_id_filter),
+            AttemptLog.objects.filter(dataset_id_filter),
+            ExamAttemptLog.objects.filter(dataset_id_filter),
+            ExamLog.objects.filter(dataset_id_filter),
+            MasteryLog.objects.filter(dataset_id_filter),
+            UserSessionLog.objects.filter(dataset_id_filter),
+        ],
+    )
+
+
+def _get_morango_models(dataset_id):
+    querysets = [DatabaseMaxCounter.objects.filter(partition__startswith=dataset_id)]
```

Comment:
There's a case we should double-check here (I know @bjester and I dealt with it somewhere) around the records created on this device itself (which will be a common case). I believe it saves a DatabaseMaxCounter with empty partition for its own instance_id, given that the local device, by definition (except with this sort of removal), has all of the data that it produced itself. I believe we dealt with that by converting the ones with empty partitions into ones with the dataset prefixes for other facilities on the device, or similar. Will come back to this.

Reply:
I don't recall that

Reply:
I looked into this a bit further. The scenario of concern relates to when we serialize from the app models into the Store, and then update the DatabaseMaxCounter values for our own instance ([here](https://github.com/learningequality/morango/blob/master/morango/sync/operations.py#L258)). We used to do unfiltered serialization (serialize everything, even if it doesn't match the filter), which means that linked code would set the DatabaseMaxCounter for its own instance_id to `""`, basically saying (correctly) "I have all records, across all partitions, for my own instance" (because, without removal, that's intrinsically true -- if data originated on this instance, then this instance automatically already has it all). If we then removed data that had originated on that instance, but left the `""` counter, it would still think it had everything for that instance, so if we tried to sync data for the removed facility back onto the device, it wouldn't get pulled in.

In practice:
- We shifted in Morango (at some point) to having serialization be filter-specific, rather than serializing everything unconditionally prior to each sync. So the line of code linked above will basically never be hit anymore, because `filter` will always have a value set.
- However, devices that had synced on previous versions may still have those old DatabaseMaxCounter records around. For instance, there are 8 of them on Kolibri Data Portal. So those devices could be impacted if we don't do something here.

What I would propose is that as a separate operation during facility deletion, we get the list of `DatabaseMaxCounter.objects.filter(partition="")`, create `[]`

```
for dmc in DatabaseMaxCounter.objects.filter(partition=""):
    for dataset in FacilityDataset.objects.all():
        DatabaseMaxCounter.objects.get_or_create(instance_id=dmc.instance_id, partition=dataset.id, counter=dmc.counter)
    dmc.delete()
```

Reply:
This is now implemented.